README.md
# termslang


ChangeLog.md
# Changelog for termslang

## Unreleased changes


test.asm
INIT
JUMP foo
JUMPDEST bar
PUSH 0x99
PUSH 0x88
MUL
PUSH 0x6677
SSTORE
JUMPDEST foo
PUSH 0x88
SLOAD
JUMPI bar
STOP


bank.asm
            ; Write 0xe1fffcc4923d04b559f4d29a8bfc6cda04eb5b0d3c460751c2402c5c5cc9109c to record DEPOSIT EVENT
            ; write ^1 to record ^2
PUSH 0xe1fffcc4923d04b559f4d29a8bfc6cda04eb5b0d3c460751c2402c5c5cc9109c
PUSH 0x20
SSTORE
            ; Write 0x884edad9ce6fa2440d8a54cc123490eb96d2768479d49ff9c7366125a9424364 to record WITHDRAW EVENT
            ; write ^1 to record ^2
PUSH 0x884edad9ce6fa2440d8a54cc123490eb96d2768479d49ff9c7366125a9424364
PUSH 0x40
SSTORE
            ; Write 0x7ae187a0c04cf55b655ca83fa11d37854c882bf1fdcb588469b414731bb0e05a to record WITHDRAWFAILED EVENT
            ; write ^1 to record ^2
PUSH 0x7ae187a0c04cf55b655ca83fa11d37854c882bf1fdcb588469b414731bb0e05a
PUSH 0x60
SSTORE
            ; CONDITIONS
            ; conditions
INIT
            ; Payable fallback
            ; payable fallback
FALLBACK ()
            ; push REVENUE CONSTANT
            ; push ^1
CALLVALUE
            ; false 1001
            ; false ^1
CONDITIONNOT 1001
            ; stop
            ; stop
STOP
            ; end 1001
            ; end ^1
CONDITIONEND 1001
            ; Find index INDEX for address CALLER CONSTANT packed in USER #0 of bytes12 , address
            ; find index ^1 for address ^2 packed in ^3 of bytes12 , address
PUSH 0x00
PUSH 0x20
JUMPDEST TAGloop_MATCH_NUMBER9
DUP1
PUSH 0x010000
ADD
SLOAD
DUP1
DUP1
ISZERO
JUMPI TAGend_of_seq_MATCH_NUMBER9
PUSH 0x60
PUSH 0x02
EXP
DUP1
SWAP2
MUL
DIV
DUP1
CALLER
EQ
JUMPI TAGaddress_match_MATCH_NUMBER9
EQ
JUMPI TAGset_replacespot_MATCH_NUMBER9
PUSH 0x20
ADD
JUMP TAGloop_MATCH_NUMBER9
JUMPDEST TAGset_replacespot_MATCH_NUMBER9
SWAP1
POP
DUP1
PUSH 0x20
ADD
JUMP TAGloop_MATCH_NUMBER9
JUMPDEST TAGend_of_seq_MATCH_NUMBER9
POP
POP
SWAP1
DUP1
JUMPI TAGfinish_MATCH_NUMBER9
SWAP1
JUMP TAGfinish_MATCH_NUMBER9
JUMPDEST TAGaddress_match_MATCH_NUMBER9
POP
POP
JUMPDEST TAGfinish_MATCH_NUMBER9
PUSH 0x20
SWAP1
DIV
PUSH 0x80
MSTORE
            ; Read record USER #INDEX into bytes12 BALANCE , address void
            ; read record ^1 into bytes12 ^2 , address void
PUSH 0x80
MLOAD
PUSH 0x20
MUL
PUSH 0x010000
ADD
SLOAD
PUSH 0xa0
PUSH 0x02
EXP
SWAP1
DIV
PUSH 0xa0
MSTORE
            ; Increment BALANCE by REVENUE CONSTANT
            ; increment ^1 by ^2
CALLVALUE
PUSH 0xa0
MLOAD
ADD
PUSH 0xa0
MSTORE
            ; Write packed bytes12 BALANCE , address CALLER CONSTANT to record USER #INDEX
            ; write packed bytes12 ^1 , address ^2 to record ^3
PUSH 0xa0
MLOAD
PUSH 0xa0
PUSH 0x02
EXP
MUL
CALLER
OR
PUSH 0x80
MLOAD
PUSH 0x20
MUL
PUSH 0x010000
ADD
SSTORE
            ; Log DEPOSIT EVENT with topics CALLER CONSTANT , REVENUE CONSTANT
            ; log ^1 with topics ^2 , ^3
CALLVALUE
CALLER
PUSH 0x20
SLOAD
PUSH 0x20
PUSH 0x20
LOG3
            ; Return
            ; return
STOP
            ; jumpdest withdraw()
            ; jumpdest ^1
JUMPDEST withdraw()
            ; Find index INDEX of address CALLER CONSTANT packed in USER #0 of bytes12 , address
            ; find index ^1 of address ^2 packed in ^3 of bytes12 , address
PUSH 0x20
DUP1
DUP1
JUMPDEST TAGloop_MATCH_NUMBER16
PUSH 0x010000
ADD
SLOAD
DUP1
ISZERO
JUMPI TAGend0_MATCH_NUMBER16
PUSH 0x60
PUSH 0x02
EXP
DUP1
SWAP2
MUL
DIV
CALLER
EQ
JUMPI TAGend_MATCH_NUMBER16
PUSH 0x20
ADD
DUP1
JUMP TAGloop_MATCH_NUMBER16
JUMPDEST TAGend0_MATCH_NUMBER16
POP
POP
POP
PUSH 0x00
JUMPDEST TAGend_MATCH_NUMBER16
DIV
PUSH 0x80
MSTORE
            ; Read record USER #INDEX into bytes12 BALANCE , address void
            ; read record ^1 into bytes12 ^2 , address void
PUSH 0x80
MLOAD
PUSH 0x20
MUL
PUSH 0x010000
ADD
SLOAD
PUSH 0xa0
PUSH 0x02
EXP
SWAP1
DIV
PUSH 0xa0
MSTORE
            ; push BALANCE
            ; push ^1
PUSH 0xa0
MLOAD
            ; false 1002
            ; false ^1
CONDITIONNOT 1002
            ; stop
            ; stop
STOP
            ; end 1002
            ; end ^1
CONDITIONEND 1002
            ; Write packed bytes12 0 , address CALLER CONSTANT to record USER #INDEX
            ; write packed bytes12 ^1 , address ^2 to record ^3
PUSH 0x00
PUSH 0xa0
PUSH 0x02
EXP
MUL
CALLER
OR
PUSH 0x80
MLOAD
PUSH 0x20
MUL
PUSH 0x010000
ADD
SSTORE
            ; Send BALANCE to CALLER CONSTANT or see 2_2 on failure
            ; send ^1 to ^2 or see ^3 on failure
PUSH 0xa0
MLOAD
ISZERO
JUMPI TAG2_2
PUSH 0x00
PUSH 0x00
PUSH 0x00
PUSH 0x00
PUSH 0xa0
MLOAD
CALLER
PUSH 0x64
GAS
SUB
CALL
ISZERO
JUMPI TAG2_2
            ; Log WITHDRAW EVENT with topics CALLER CONSTANT , BALANCE
            ; log ^1 with topics ^2 , ^3
PUSH 0xa0
MLOAD
CALLER
PUSH 0x40
SLOAD
PUSH 0x20
PUSH 0x20
LOG3
            ; Stop
            ; stop
STOP
            ; jumpdest 2_2
            ; jumpdest ^1
JUMPDEST TAG2_2
            ; Write packed bytes12 BALANCE , address CALLER CONSTANT to record USER #INDEX
            ; write packed bytes12 ^1 , address ^2 to record ^3
PUSH 0xa0
MLOAD
PUSH 0xa0
PUSH 0x02
EXP
MUL
CALLER
OR
PUSH 0x80
MLOAD
PUSH 0x20
MUL
PUSH 0x010000
ADD
SSTORE
            ; Log WITHDRAWFAILED EVENT with topic CALLER CONSTANT
            ; log ^1 with topic ^2
CALLER
PUSH 0x60
SLOAD
PUSH 0x20
DUP1
LOG2
            ; Return
            ; return
STOP
            ; jumpdest balance()
            ; jumpdest ^1
JUMPDEST balance()
            ; Find index INDEX of address CALLER CONSTANT packed in USER #0 of bytes12 , address
            ; find index ^1 of address ^2 packed in ^3 of bytes12 , address
PUSH 0x20
DUP1
DUP1
JUMPDEST TAGloop_MATCH_NUMBER31
PUSH 0x010000
ADD
SLOAD
DUP1
ISZERO
JUMPI TAGend0_MATCH_NUMBER31
PUSH 0x60
PUSH 0x02
EXP
DUP1
SWAP2
MUL
DIV
CALLER
EQ
JUMPI TAGend_MATCH_NUMBER31
PUSH 0x20
ADD
DUP1
JUMP TAGloop_MATCH_NUMBER31
JUMPDEST TAGend0_MATCH_NUMBER31
POP
POP
POP
PUSH 0x00
JUMPDEST TAGend_MATCH_NUMBER31
DIV
PUSH 0x80
MSTORE
            ; Read record USER #INDEX into bytes12 BALANCE , address void
            ; read record ^1 into bytes12 ^2 , address void
PUSH 0x80
MLOAD
PUSH 0x20
MUL
PUSH 0x010000
ADD
SLOAD
PUSH 0xa0
PUSH 0x02
EXP
SWAP1
DIV
PUSH 0xa0
MSTORE
            ; Return uint256 at BALANCE
            ; return uint256 at ^1
PUSH 0x20
PUSH 0xa0
RETURN
            ; jumpdest users(uint256 index)
            ; jumpdest ^1
JUMPDEST users(uint256)
            ; Let INDEX get argument 1
            ; let ^1 get argument 1
PUSH 0x04
CALLDATALOAD
PUSH 0x80
MSTORE
            ; Read record USER #INDEX into bytes12 BALANCE , address USER
            ; read record ^1 into bytes12 ^2 , address ^3
PUSH 0x80
MLOAD
PUSH 0x20
MUL
PUSH 0x010000
ADD
SLOAD
DUP1
PUSH 0xa0
PUSH 0x02
EXP
SWAP1
DIV
PUSH 0xa0
MSTORE
PUSH 0x60
PUSH 0x02
EXP
DUP1
SWAP2
MUL
DIV
PUSH 0xc0
MSTORE
            ; Return address at USER
            ; return address at ^1
PUSH 0x14
PUSH 0xc0
PUSH 0x0C
ADD  ; add
RETURN;return
            ; jumpdest funds(uint256 index)
            ; jumpdest ^1


JUMPDEST funds(uint256) ;funds
            ; Let INDEX get argument 1
            ; let ^1 get argument 1
PUSH 0x04
CALLDATALOAD
PUSH 0x80
MSTORE
            ; Read record USER #INDEX into bytes12 BALANCE , address void
            ; read record ^1 into bytes12 ^2 , address void
PUSH 0x80
MLOAD; mload
PUSH 0x20
MUL
PUSH 0x010000
ADD
SLOAD
PUSH 0xa0
PUSH 0x02
EXP
SWAP1
DIV
PUSH 0xa0
MSTORE
            ; Return uint256 at BALANCE
            ; return uint256 at ^1
PUSH 0x20
PUSH 0xa0
RETURN
            ; jumpdest 0
            ; jumpdest ^1
JUMPDEST TAG0
            ; return
            ; return
STOP;


token.asm
            ; Write 0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef to record TRANSFER EVENT
            ; write ^1 to record ^2
PUSH 0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef
PUSH 0x20
SSTORE
            ; Write 0x8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925 to record APPROVAL EVENT
            ; write ^1 to record ^2
PUSH 0x8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925
PUSH 0x40
SSTORE
            ; Write 1000 to record SUPPLY
            ; write ^1 to record ^2
PUSH 0x03e8
PUSH 0x60
SSTORE
            ; Write 1000 to record 1 of ORIGIN CONSTANT
            ; write ^1 to record ^2 of ^3
PUSH 0x03e8
ORIGIN
PUSH 0x01
PUSH 0x20
MUL
ADD
SSTORE
            ; Write ORIGIN CONSTANT to record OWNER
            ; write ^1 to record ^2
ORIGIN
PUSH 0x80
SSTORE
            ; Write string 0x5465726d7320746f6b656e to record NAME #0
            ; write string ^1 to record ^2
PUSH 0x010000
SSTORESEQ 0x5465726d7320746f6b656e
            ; Write string 0x5454 to record SYMBOL #0
            ; write string ^1 to record ^2
PUSH 0x020000
SSTORESEQ 0x5454
            ; Write 18 to record DECIMALS
            ; write ^1 to record ^2
PUSH 0x12
PUSH 0xa0
SSTORE
            ; CONDITIONS
            ; conditions
INIT
            ; Payable fallback
            ; payable fallback
FALLBACK ()
            ; Return
            ; return
STOP
            ; jumpdest transfer(address to, uint256 value)
            ; jumpdest ^1
JUMPDEST transfer(address,uint256)
            ; Let TO get argument 1
            ; let ^1 get argument 1
PUSH 0x04
CALLDATALOAD
PUSH 0xc0
MSTORE
            ; Let VALUE get argument 2
            ; let ^1 get argument 2
PUSH 0x24
CALLDATALOAD
PUSH 0xe0
MSTORE
            ; Let FROM = CALLER CONSTANT
            ; let ^1 = ^2
CALLER
PUSH 0x0100
MSTORE
            ; Apply procedure "transfer"
            ; apply procedure ^1
REFJUMP TAG_TRANSFER_PROCEDURE
            ; Return
            ; return
STOP
            ; jumpdest transferFrom(address from, address to, uint256 value)
            ; jumpdest ^1
JUMPDEST transferFrom(address,address,uint256)
            ; Let FROM get argument 1
            ; let ^1 get argument 1
PUSH 0x04
CALLDATALOAD
PUSH 0x0100
MSTORE
            ; Let TO get argument 2
            ; let ^1 get argument 2
PUSH 0x24
CALLDATALOAD
PUSH 0xc0
MSTORE
            ; Let VALUE get argument 3
            ; let ^1 get argument 3
PUSH 0x44
CALLDATALOAD
PUSH 0xe0
MSTORE
            ; Let TMP = FROM - CALLER CONSTANT
            ; let ^1 = ^2 - ^3
CALLER
PUSH 0x0100
MLOAD
SUB
PUSH 0x0120
MSTORE
            ; Let REMAINING read record TMP
            ; let ^1 read record ^2
PUSH 0x0120
MLOAD
SLOAD
PUSH 0x0140
MSTORE
            ; push REMAINING
            ; push ^1
PUSH 0x0140
MLOAD
            ; push VALUE
            ; push ^1
PUSH 0xe0
MLOAD
            ; comp <=
            ; comp <=
GT
ISZERO
            ; false 1001
            ; false ^1
CONDITIONNOT 1001
            ; stop
            ; stop
STOP
            ; end 1001
            ; end ^1
CONDITIONEND 1001
            ; Decrement REMAINING by VALUE
            ; decrement ^1 by ^2
PUSH 0xe0
MLOAD
PUSH 0x0140
MLOAD
SUB
PUSH 0x0140
MSTORE
            ; Write REMAINING to record at TMP
            ; write ^1 to record at ^2
PUSH 0x0140
MLOAD
PUSH 0x0120
MLOAD
SSTORE
            ; Apply procedure "transfer"
            ; apply procedure ^1
REFJUMP TAG_TRANSFER_PROCEDURE
            ; Return
            ; return
STOP
            ; jumpdest approve(address spender, uint256 value)
            ; jumpdest ^1
JUMPDEST approve(address,uint256)
            ; Let SPENDER get argument 1
            ; let ^1 get argument 1
PUSH 0x04
CALLDATALOAD
PUSH 0x0160
MSTORE
            ; Let VALUE get argument 2
            ; let ^1 get argument 2
PUSH 0x24
CALLDATALOAD
PUSH 0xe0
MSTORE
            ; Let TMP = CALLER CONSTANT - SPENDER
            ; let ^1 = ^2 - ^3
PUSH 0x0160
MLOAD
CALLER
SUB
PUSH 0x0120
MSTORE
            ; Write VALUE to record at TMP
            ; write ^1 to record at ^2
PUSH 0xe0
MLOAD
PUSH 0x0120
MLOAD
SSTORE
            ; Log APPROVAL EVENT with topics CALLER CONSTANT , SPENDER , VALUE
            ; log ^1 with topics ^2 , ^3 , ^4
PUSH 0xe0
MLOAD
PUSH 0x0160
MLOAD
CALLER
PUSH 0x40
SLOAD
PUSH1 0x20
DUP1
LOG4
            ; Return
            ; return
STOP
            ; jumpdest mint(uint256 amount, address to)
            ; jumpdest ^1
JUMPDEST mint(uint256,address)
            ; Let AMOUNT get argument 1
            ; let ^1 get argument 1
PUSH 0x04
CALLDATALOAD
PUSH 0x0180
MSTORE
            ; Let TO get argument 2
            ; let ^1 get argument 2
PUSH 0x24
CALLDATALOAD
PUSH 0xc0
MSTORE
            ; push TO
            ; push ^1
PUSH 0xc0
MLOAD
            ; false 1002
            ; false ^1
CONDITIONNOT 1002
            ; let TO = CALLER CONSTANT
            ; let ^1 = ^2
CALLER
PUSH 0xc0
MSTORE
            ; end 1002
            ; end ^1
CONDITIONEND 1002
            ; Let BALANCE read record 1 of TO
            ; let ^1 read record ^2 of ^3
PUSH 0xc0
MLOAD
PUSH 0x01
PUSH 0x20
MUL
ADD
SLOAD
PUSH 0x01a0
MSTORE
            ; Grab record SUPPLY
            ; grab record ^1
PUSH 0x60
DUP1
SLOAD
SWAP1
MSTORE
            ; Increment BALANCE by AMOUNT
            ; increment ^1 by ^2
PUSH 0x0180
MLOAD
PUSH 0x01a0
MLOAD
ADD
PUSH 0x01a0
MSTORE
            ; Increment SUPPLY by AMOUNT
            ; increment ^1 by ^2
PUSH 0x0180
MLOAD
PUSH 0x60
MLOAD
ADD
PUSH 0x60
MSTORE
            ; Write BALANCE to record 1 of TO
            ; write ^1 to record ^2 of ^3
PUSH 0x01a0
MLOAD
PUSH 0xc0
MLOAD
PUSH 0x01
PUSH 0x20
MUL
ADD
SSTORE
            ; Write record SUPPLY
            ; write record ^1
PUSH 0x60
MLOAD
PUSH 0x60
SSTORE
            ; Return
            ; return
STOP
            ; jumpdest balanceOf(address owner)
            ; jumpdest ^1
JUMPDEST balanceOf(address)
            ; Let OWNER get argument 1
            ; let ^1 get argument 1
PUSH 0x04
CALLDATALOAD
PUSH 0x80
MSTORE
            ; Let BALANCE read record 1 of OWNER
            ; let ^1 read record ^2 of ^3
PUSH 0x80
MLOAD
PUSH 0x01
PUSH 0x20
MUL
ADD
SLOAD
PUSH 0x01a0
MSTORE
            ; Return uint256 at BALANCE
            ; return uint256 at ^1
PUSH 0x20
PUSH 0x01a0
RETURN
            ; jumpdest allowance(address from, address to)
            ; jumpdest ^1
JUMPDEST allowance(address,address)
            ; Let FROM get argument 1
            ; let ^1 get argument 1
PUSH 0x04
CALLDATALOAD
PUSH 0x0100
MSTORE
            ; Let TO get argument 2
            ; let ^1 get argument 2
PUSH 0x24
CALLDATALOAD
PUSH 0xc0
MSTORE
            ; Let TMP = FROM - TO
            ; let ^1 = ^2 - ^3
PUSH 0xc0
MLOAD
PUSH 0x0100
MLOAD
SUB
PUSH 0x0120
MSTORE
            ; Let REMAINING read record TMP
            ; let ^1 read record ^2
PUSH 0x0120
MLOAD
SLOAD
PUSH 0x0140
MSTORE
            ; Return uint256 at REMAINING
            ; return uint256 at ^1
PUSH 0x20
PUSH 0x0140
RETURN
            ; jumpdest name()
            ; jumpdest ^1
JUMPDEST name()
            ; Grab string NAME #0
            ; grab string ^1
PUSH 0x010000
JUMPDEST TAGloop_MATCH_NUMBER65
DUP1
DUP1
SLOAD
SWAP1
MSTORE
PUSH 0x20
ADD
DUP1
SLOAD
JUMPI TAGloop_MATCH_NUMBER65
            ; Return string at NAME #0
            ; return string at ^1
PUSH 0x010000
PUSH 0x00
MSTORE
PUSH 0x010000
DUP1
MLOAD
ADD
PUSH 0x20
ADD
PUSH 0x00
RETURN
            ; jumpdest symbol()
            ; jumpdest ^1
JUMPDEST symbol()
            ; Grab string SYMBOL #0
            ; grab string ^1
PUSH 0x020000
JUMPDEST TAGloop_MATCH_NUMBER68
DUP1
DUP1
SLOAD
SWAP1
MSTORE
PUSH 0x20
ADD
DUP1
SLOAD
JUMPI TAGloop_MATCH_NUMBER68
            ; Return string at SYMBOL #0
            ; return string at ^1
PUSH 0x020000
PUSH 0x00
MSTORE
PUSH 0x020000
DUP1
MLOAD
ADD
PUSH 0x20
ADD
PUSH 0x00
RETURN
            ; jumpdest decimals()
            ; jumpdest ^1
JUMPDEST decimals()
            ; Grab record DECIMALS
            ; grab record ^1
PUSH 0xa0
DUP1
SLOAD
SWAP1
MSTORE
            ; Return uint256 at DECIMALS
            ; return uint256 at ^1
PUSH 0x20
PUSH 0xa0
RETURN
            ; jumpdest totalSupply()
            ; jumpdest ^1
JUMPDEST totalSupply()
            ; Grab record SUPPLY
            ; grab record ^1
PUSH 0x60
DUP1
SLOAD
SWAP1
MSTORE
            ; Return uint256 at SUPPLY
            ; return uint256 at ^1
PUSH 0x20
PUSH 0x60
RETURN
            ; Procedure "transfer"
            ; procedure ^1
JUMPDEST TAG_TRANSFER_PROCEDURE
            ; Let FROM BALANCE read record 1 of FROM
            ; let ^1 read record ^2 of ^3
PUSH 0x0100
MLOAD
PUSH 0x01
PUSH 0x20
MUL
ADD
SLOAD
PUSH 0x01c0
MSTORE
            ; push FROM BALANCE
            ; push ^1
PUSH 0x01c0
MLOAD
            ; push VALUE
            ; push ^1
PUSH 0xe0
MLOAD
            ; comp <=
            ; comp <=
GT
ISZERO
            ; false 1003
            ; false ^1
CONDITIONNOT 1003
            ; stop
            ; stop
STOP
            ; end 1003
            ; end ^1
CONDITIONEND 1003
            ; Let TO BALANCE read record 1 of TO
            ; let ^1 read record ^2 of ^3
PUSH 0xc0
MLOAD
PUSH 0x01
PUSH 0x20
MUL
ADD
SLOAD
PUSH 0x01e0
MSTORE
            ; Decrement FROM BALANCE by VALUE
            ; decrement ^1 by ^2
PUSH 0xe0
MLOAD
PUSH 0x01c0
MLOAD
SUB
PUSH 0x01c0
MSTORE
            ; Increment TO BALANCE by VALUE
            ; increment ^1 by ^2
PUSH 0xe0
MLOAD
PUSH 0x01e0
MLOAD
ADD
PUSH 0x01e0
MSTORE
            ; Write FROM BALANCE to record 1 of FROM
            ; write ^1 to record ^2 of ^3
PUSH 0x01c0
MLOAD
PUSH 0x0100
MLOAD
PUSH 0x01
PUSH 0x20
MUL
ADD
SSTORE
            ; Write TO BALANCE to record 1 of TO
            ; write ^1 to record ^2 of ^3
PUSH 0x01e0
MLOAD
PUSH 0xc0
MLOAD
PUSH 0x01
PUSH 0x20
MUL
ADD
SSTORE
            ; Grab record TRANSFER EVENT
            ; grab record ^1
PUSH 0x20
DUP1
SLOAD
SWAP1
MSTORE
            ; Log TRANSFER EVENT with topics FROM , TO , data VALUE
            ; log ^1 with topics ^2 , ^3 , data ^4
PUSH 0xc0
MLOAD
PUSH 0x0100
MLOAD
PUSH 0x20
SLOAD
PUSH 0x20
PUSH 0xe0
LOG3
            ; Procedure end
            ; procedure end
BACKJUMP
            ; jumpdest 0
            ; jumpdest ^1
JUMPDEST TAG0
            ; return
            ; return
STOP


src/Emasm.hs
module Emasm
  ( parseAsmFileContents
  , unfoldPseudoasm
  , splitInitSection
  , computeBytecode
  , extractMethods
  , uniteBytecode
  , addLoader
  -- , printASMLine
  ) where


import Data.List
import Data.Char
import Common (keccakHash, showHex, hexFromInstr)



data ASMLine = ASMLine  -- String String
  { instr    :: String
  , operand  :: String
  , bytecode :: String
  , offset   :: Int
  }
  deriving (Eq)
instance Show ASMLine where
  show (ASMLine instr operand bytecode offset) = "(" ++ init (show instr) ++ " " ++ tail (show operand) ++ ",          " ++ show bytecode ++ ", offset " ++ show offset ++ ")"


parseAsmFileContents :: String -> [ASMLine]
parseAsmFileContents =
  map (\(x,y) -> ASMLine {instr = x, operand = dropWhile isSpace y, bytecode = "", offset = 0}) .
  map (break isSpace) . filter (not . null) . map (dropWhileEnd isSpace) .
  map (dropWhile isSpace) . map (takeWhile (/=';')) . lines




-- unfoldPush :: (String,String) -> (String,String)
-- unfoldPush (x,y)
--   | (x == "PUSH") = ("PUSH" ++ (show n), y)
--   | otherwise     = (x,y)
--   where
--     n = (length y) `div` 2 - 1


splitInitSection :: [(String,String)] -> ([(String,String)], [(String,String)])
splitInitSection arr = (\(x,y) -> (x, drop 1 y)) $ break (==("INIT","")) arr


isMethodSignature :: String -> Bool
isMethodSignature x
  | elem '(' x && last x == ')'    = True
  | otherwise                      = False


--TODO: index methods in list
extractMethods :: [(String,String)] -> [String]
extractMethods asm = [ y | (x,y) <- asm, ((x == "JUMPDEST" || x == "FALLBACK") && isMethodSignature y) ]

extractTags :: [(String,String)] -> [String]
extractTags asm = [ y | (x,y) <- asm, ((x == "JUMPDEST" || x == "FALLBACK")) ]



-- unfoldFallback :: (String,String) -> (String,String)
-- unfoldFallback (x,y)




unfoldPseudoasm :: [ASMLine] -> [ASMLine]
unfoldPseudoasm pseudoasm = pseudoasm





precomputeBytecode :: [ASMLine] -> [ASMLine]
precomputeBytecode asm = map (\x -> x {bytecode = hexFromInstr (instr x)(operand x)}) asm


computeBytecode :: [ASMLine] -> [ASMLine]
-- computeBytecode asm = precomputeBytecode asm
computeBytecode asm = resolveJumps $ precomputeBytecode asm



uniteBytecode :: [ASMLine] -> String
uniteBytecode asm = foldl (++) "" [ bytecode x | x <- asm ]
-- uniteBytecode :: [(String,String,String)] -> String
-- uniteBytecode asm = foldl (++) "" [ z | (_,_,z) <- asm , True ]


-- totalLength :: [ASMLine] -> Int
-- totalLength a = foldl (+) 0 [ length (bytecode x) | x <- a ]

computeOffsets :: [ASMLine] -> [ASMLine]
computeOffsets [] = []
computeOffsets (a:as) = let
  t = length (uniteBytecode (a:as))
  v = foldl (+) 0 [ length (bytecode x) | x <- as ]

  in a {offset = v}: computeOffsets as
  -- where
  -- oo = foldl (+) 0 (offset (a:as))
  -- oo = scanl (+) 0 . map (\x -> length (bytecode x)) $ (a:as)

{-
computeOffsets :: [ASMLine] -> [ASMLine]
computeOffsets :: [a] -> [b] -> [(a,b)]
computeOffsets []     _bs    = []
computeOffsets _as    []     = []
computeOffsets (a:as) (b:bs) = (a,b) : computeOffsets as bs
  where
  oo = scanl (+) 0 . map (\x -> length (bytecode x)) $ asm
-}


--
-- computeOffsets :: [ASMLine] -> [ASMLine]
-- -- computeOffsets asm = map (\x -> x {offset = oo}) asm
-- computeOffsets asm =
--   -- [ x {offset = y} | x <- asm , y <- oo ]
--   map (\x -> x {offset = 2}) asm
--
--   -- map (\x -> x {offset = oo}) asm
--   where
--   oo = scanl (+) 0 . map (\x -> length (bytecode x)) $ asm
-- --  foldl (+) (length (bytecode x) `div` 2)--  (\x -> x {offset = length (bytecode x) `div` 2}) asm
-- -- computeOffsets asm = map (\x -> x {offset = length (bytecode x) `div` 2}) asm

  -- l = map (\x -> length (bytecode x)) $ asm
  -- l = head . scanl (+) 0 . map (\x -> length (bytecode x)) $ asm
  -- -- in foldr (\(x,y,z) -> (x,y,z,show l)) asm ("","","","")
  -- in map (\(x,y,z) -> (x,y,z,show l)) asm


resolveJumps :: [ASMLine] -> [ASMLine]
resolveJumps asm = computeOffsets asm










addLoader :: String -> String
addLoader bytecode = let
  initial = "hdfhsbhb"
  codesize = length bytecode `div` 2
  codepush = if codesize > 0xff then "61" else "60"
  a = initial ++ codepush ++ showHex codesize ++ "80"
  codeoffset1 = length a `div` 2 + 8
  initpush = if codeoffset > 0xff then "61" else "60"
  codeoffset = if codeoffset1 > 0xff then codeoffset1 + 1 else codeoffset1
  b = a ++ initpush ++ showHex codeoffset ++ "6000396000f3" ++ bytecode
  in b


src/Common.hs
module Common
  ( keccakHash
  , showHex
  , hexFromInstr
  ) where


import Data.Char
import Numeric (showIntAtBase)

import Data.Text.Encoding   (encodeUtf8)
import Data.Text.Conversions
import Crypto.Hash          (hash, Keccak_256, Digest)
keccakHash :: String -> String
keccakHash x = show (hash (encodeUtf8 (convertText (x))) :: Digest Keccak_256)


showHex :: (Integral a, Show a) => a -> String
showHex x = if odd (length s) then '0':s else s
  where s = (showIntAtBase 16 intToDigit) x ""


hexFromInstr :: String -> String -> String
hexFromInstr instr operand =
  case instr of
    "PUSH" -> showHex (0x5f + (length operand) `div` 2 - 1) ++ drop 2 operand
    _ -> yy !! 0
  where
    yy = [ y | (x,y) <- l , x == instr ] ++ [""]
    l = [
      ("STOP", "00"),
      ("ADD", "01"),
      ("MUL", "02"),
      ("SUB", "03"),
      ("DIV", "04"),
      ("SDIV", "05"),
      ("MOD", "06"),
      ("SMOD", "07"),
      ("ADDMOD", "08"),
      ("MULMOD", "09"),
      ("EXP", "0a"),
      ("SIGNEXTEND", "0b"),

      ("LT", "10"),
      ("GT", "11"),
      ("SLT", "12"),
      ("SGT", "13"),
      ("EQ", "14"),
      ("ISZERO", "15"),
      ("AND", "16"),
      ("OR", "17"),
      ("XOR", "18"),
      ("NOT", "19"),
      ("BYTE", "1a"),

      ("SHA3", "20"),

      ("ADDRESS", "30"),
      ("BALANCE", "31"),
      ("ORIGIN", "32"),
      ("CALLER", "33"),
      ("CALLVALUE", "34"),
      ("CALLDATALOAD", "35"),
      ("CALLDATASIZE", "36"),
      ("CALLDATACOPY", "37"),
      ("CODESIZE", "38"),
      ("CODECOPY", "39"),
      ("GASPRICE", "3a"),
      ("EXTCODESIZE", "3b"),
      ("EXTCODECOPY", "3c"),
      ("RETURNDATASIZE", "3d"),
      ("RETURNDATACOPY", "3e"),

      ("BLOCKHASH", "40"),
      ("COINBASE", "41"),
      ("TIMESTAMP", "42"),
      ("NUMBER", "43"),
      ("DIFFICULTY", "44"),
      ("GASLIMIT", "45"),

      ("POP", "50"),
      ("MLOAD", "51"),
      ("MSTORE", "52"),
      ("MSTORE8", "53"),
      ("SLOAD", "54"),
      ("SSTORE", "55"),
      ("JUMP", "56"),
      ("JUMPI", "57"),
      ("PC", "58"),
      ("MSIZE", "59"),
      ("GAS", "5a"),
      ("JUMPDEST", "5b"),

      ("DUP1", "80"),
      ("DUP2", "81"),
      ("DUP3", "82"),
      ("DUP4", "83"),
      ("DUP5", "84"),
      ("DUP6", "85"),
      ("DUP7", "86"),
      ("DUP8", "87"),
      ("DUP9", "88"),
      ("DUP10", "89"),
      ("DUP11", "8a"),
      ("DUP12", "8b"),
      ("DUP13", "8c"),
      ("DUP14", "8d"),
      ("DUP15", "8e"),
      ("DUP16", "8f"),

      ("SWAP1", "90"),
      ("SWAP2", "91"),
      ("SWAP3", "92"),
      ("SWAP4", "93"),
      ("SWAP5", "94"),
      ("SWAP6", "95"),
      ("SWAP7", "96"),
      ("SWAP8", "97"),
      ("SWAP9", "98"),
      ("SWAP10", "99"),
      ("SWAP11", "9a"),
      ("SWAP12", "9b"),
      ("SWAP13", "9c"),
      ("SWAP14", "9d"),
      ("SWAP15", "9e"),
      ("SWAP16", "9f"),

      ("LOG0", "a0"),
      ("LOG1", "a1"),
      ("LOG2", "a2"),
      ("LOG3", "a3"),
      ("LOG4", "a4"),
-- opcodes е1, е2, е3 deprecated
      ("CREATE", "f0"),
      ("CALL", "f1"),
      ("CALLCODE", "f2"),
      ("RETURN", "f3"),
      ("DELEGATECALL", "f4"),
      ("CALLBLACKBOX", "f5"),

      ("STATICCALL", "fa"),
      ("REVERT", "fd"),
      ("INVALID", "fe"),

      ("SELFDESTRUCT", "ff"),
      ("SUICIDE", "ff")
      ]
